Support for pca9634 led driver.

mzagrabski opened the pull request:
Little cleanup of pca963x (and similar) drivers, while adding support for pca9634.

Commit:
drivers: led: pca9633: rename driver into pca_common

The PCA9633 is not only member of device family.
Rename driver to make a room for expansion.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Mani-Sadhasivam commented:
@mzagrabski There is also a PR open for the PCA9956 chip. Looking at the existing driver and the open PRs for these LED family, it looks to me that there could a common `pca_common.c` file handling the settings and individual drivers containing the config for each LEDs.

What do you think?

mzagrabski commented:
@Mani-Sadhasivam the pca9956 is similar enough, to be easily adopted here. I renamed driver `pca9633.c` into `pca963x.c`, so i will just rename it into `pca_common.c`, and add pca9956 support on top.

mbolivar-nordic commented on drivers/led/pca963x.c:
```diff
@@ -0,0 +1,265 @@
+		.grppwm = regs ## _GRPPWM,							\
+		.grpfreq = regs ## _GRPFREQ,							\
+		.ledout = regs ## _LEDOUT,							\
+		.external_driver = DT_PROP_OR(node_id, external_driver, false),			\
```
Just use DT_PROP(node_id, external_driver). DT_PROP() does the right thing for boolean properties. It knows that a missing boolean property is `false`.

mbolivar-nordic commented on dts/bindings/led/nxp,pca9634.yaml:
```diff
@@ -0,0 +1,10 @@
+include: i2c-device.yaml
+
+properties:
+  external-driver:
+    type: boolean
+    required: false
```
Please add a `description:` with an explanation of what setting this to `true` means.

mzagrabski replied:
Done.

mbolivar-nordic commented on drivers/led/pca963x.c:
```diff
@@ -0,0 +1,265 @@
+};
+
+#define PCA963x_DEVICE(node_id, prefix, regs, params)						\
+	static const struct pca963x_config _CONCAT(prefix ## _config_, DT_DEP_ORD(node_id)) = {	\
```
Nice trick.

mbolivar-nordic reviewed (changes requested):
A couple of minor things on the DT/driver side from me.

Commit:
drivers: led: pca_common: enable driver when device is present

Set CONFIG_PCA_COMMON default value to y if any compatible
device is present in the devicetree.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
drivers: led: pca_common: support external driver mode

PCA9634 have dedicated mode for when led is driven
via external mosfet. Provide boolean property
in devicetree that enable this mode.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
drivers: led: pca_common: move register numbers to config

Make driver more generic by moving register numbers to config.
Assume that every LEDOUT register can handle up to 4 leds.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
drivers: led: pca_common: move parameters to config

Move driver parameters such as min/max period and min/max brightness
to config, as they are constant in runtime and depend on chip.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
drivers: led: pca_common: add support for pca9634

Add support for pca9634 led driver which is extension
of pca9633 with support for 8 channels.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
drivers: led: pca_common: merge driver for tlc59108

As TI TLC59108 is almost identical device to PCA9634
merge drivers together.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
drivers: led: pca_common: add support for pca9956

Add support for pca9956 led driver with support for 24 channels.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
drivers: led: pca_common: allow to set default iref for pca9956

PCA9956 allow setting default current for all channels,
enable this feature via devicetree and init.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

Commit:
tests: drivers: build_all: Add a test build of LED driver

Move existing led driver test nodes from sensor set to led
set. Complete test set with previously added devices pca_common devices.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

mniestroj commented on dts/bindings/led/nxp,pca9956.yaml:
```diff
@@ -6,3 +6,12 @@ description: NXP PCA9956 LED
+properties:
+    default-iref:
+        type: int
+        default: 255
```
Not sure, but doesn't this default value of `255` always override `0` used in driver: `.default_iref = DT_PROP_OR(node_id, default_iref, 0),`?

mzagrabski replied:
You are right, as property has default value, I removed `DT_PROP_OR`, it is COND_CODE anyway, so I assume it if safe enough.

mzagrabski replied:
Ok, i remember now, i was thinking about exotic scenario where you have multiple different PCA chips in design, so `.default_iref` of config struct will be compiled and filled by macro, but non always will be present in dts, as it is part only of pca9956 chip, not the rest of them, in that case i wanted to `zero` this field.

mbolivar-nordic replied:
You may consider removing the default if that is appropriate.

If you need to keep it, you must justify why the choice of the default value was made **in the description in the binding** (i.e. GitHub comments with a justification **do not count**).

Otherwise bindings LGTM, thank you for your patience.

mzagrabski replied:
I have removed this default value, and remade driver so skipping property in dts will just ignore it. Thanks.

simonguinot commented on drivers/led/Kconfig.pca_common:
```diff
@@ -17,3 +17,13 @@ config PCA_COMMON
+	bool "Set external driver on startup"
+	depends on $(dt_compat_enabled,$(DT_COMPAT_NXP_PCA9634))
+	help
+	  Set external driver option, configured in dts, on startup.
```
If the external-driver property is actually defined in DT, is there a reason to allow to disable external driver support from Kconfig ? If no, then maybe this option is not needed at all.

mzagrabski replied:
I have removed Kconfig settings and replaced them it by `DT_PROP_ANY` macro.

simonguinot commented on drivers/led/Kconfig.pca_common:
```diff
@@ -29,4 +29,10 @@ config PCA_COMMON_EXTERNAL_DRIVER
 
+config PCA_COMMON_DEFAULT_IREF
+	bool "Set default iref parameter on startup"
+	depends on $(dt_compat_enabled,$(DT_COMPAT_NXP_PCA9956))
```
Again, is there a reason to allow to enable/disable this if the property is defined in DT ?

mzagrabski replied:
I have removed it and replaced by `DT_ANY_NODE_HAS_PROP` macro.

simonguinot reviewed (commented):
Hi @mzagrabski,

It is really a nice work and I don't have much to say.

Please see below for few comments and questions.

mbolivar-nordic commented on tests/drivers/build_all/led/app.overlay:
```diff
@@ -31,5 +31,40 @@
+			#size-cells = <0>;
+			compatible = "vnd,i2c";
+			reg = <0x11112222 0x1000>;
+			label = "TEST_I2C_CTLR";
```
Please remove these label properties when you get a chance. People are in the middle of a treewide cleanup to remove them.

Commit:
samples: drivers: led: Rename pca9633 into pca_common

As driver got renamed, rename sample as well.
Add board supported by driver as overlays.

Signed-off-by: Maciej Zagrabski <maciej.zagrabski@grinn-global.com>

mzagrabski commented on drivers/led/Kconfig.pca9633:
```diff
@@ -1,11 +0,0 @@
-# Copyright (c) 2018 Savoir-Faire Linux
-# SPDX-License-Identifier: Apache-2.0
-
-config PCA9633
```
Thanks for noticing! I have fixed that.

mzagrabski commented on drivers/led/Kconfig.pca_common:
```diff
@@ -3,19 +3,22 @@
 	default y if $(dt_compat_enabled,$(DT_COMPAT_NXP_PCA9633))
 	default y if $(dt_compat_enabled,$(DT_COMPAT_NXP_PCA9634))
+	default y if $(dt_compat_enabled,$(DT_COMPAT_NXP_PCA9956))
 	default y if $(dt_compat_enabled,$(DT_COMPAT_TI_TLC59108))
```
I have renamed sample, and added overlays for all supported devices.

simonguinot commented on drivers/led/pca_common.c:
```diff
@@ -0,0 +1,314 @@
+
+#define _DT_ANY_NODE_HAS_PROP_OR(id, compat, prop) DT_NODE_HAS_PROP(DT_INST(id, compat), prop) ||
+#define DT_ANY_NODE_HAS_PROP(compat, prop) \
+	DT_FOREACH_STATUS_OKAY_VARGS(compat, _DT_ANY_NODE_HAS_PROP_OR, compat, prop) 0
```
Well... it is quite something :) I suppose because `external-driver` can be false, then we can't use `DT_ANY_NODE_HAS_PROP` with it.... How many bytes of memory are we saving with this trick ? I'm not sure it's worth the readability we give. But the code looks correct.

simonguinot reviewed (approved):
Hi @mzagrabski,

The driver is quite tricky to read/review but I guess it is the price to pay for sharing it between multiple devices.

Anyway, it is looking good to me ! Thanks for your great work.

mbolivar-nordic commented on drivers/led/pca_common.c:
```diff
@@ -20,6 +20,10 @@ LOG_MODULE_REGISTER(pca_common);
 #include "led_context.h"
 
+#define _DT_PROP_ANY_OR(id, compat, prop) DT_PROP(DT_INST(id, compat), prop) ||
+#define DT_PROP_ANY(compat, prop) \
```
The `DT_` namespace is reserved for zephyr/devicetree.h and its sub-headers. Please choose a different name.

mbolivar-nordic commented on drivers/led/pca_common.c:
```diff
@@ -60,6 +66,12 @@ LOG_MODULE_REGISTER(pca_common);
+#define PCA9956_MIN_PERIOD	67U
+#define PCA9956_MAX_PERIOD	16776U
+#define PCA9956_MIN_BRIGHTNESS	0U
+#define PCA9956_MAX_BRIGHTNESS	100U
```
It seems like you could save a little more ROM by initializing one `struct led_data` per family, and pointing to it from each instance:

```
struct pca_common_config { /* ... */ const struct led_data *led_data; /* ... */ };
```

But this is not a blocking comment, please feel free to disregard this idea or defer it to a later pull request.

mbolivar-nordic commented on drivers/led/pca_common.c:
```diff
@@ -24,6 +24,10 @@ LOG_MODULE_REGISTER(pca_common);
 	DT_FOREACH_STATUS_OKAY_VARGS(compat, _DT_PROP_ANY_OR, compat, prop) 0
 
+#define _DT_ANY_NODE_HAS_PROP_OR(id, compat, prop) DT_NODE_HAS_PROP(DT_INST(id, compat), prop) ||
+#define DT_ANY_NODE_HAS_PROP(compat, prop) \
```
Please pick a different name outside of the reserved namespace.

mbolivar-nordic reviewed (changes requested):
Very nice work. A couple of minor issues, otherwise looks excellent to me.

simonguinot commented:
Hi @mzagrabski,
What are your intentions regarding this PR ? There is only a few changes needed to make it ready to merge. It would be a shame to give up now.

mzagrabski commented:
@simonguinot Lack of time stalled this, as always. I am afraid it require little bump after 3.2 release. I will try to push it this weekend.

simonguinot commented:
Hi @mzagrabski,
What are your intentions regarding this PR ? There is only a few changes needed to make it ready to merge. It would be a shame to give up now.

> @simonguinot Lack of time stalled this, as always. I am afraid it require little bump after 3.2 release. I will try to push it this weekend.

ping :)
